clients/pkg/logentry/stages/limit.go

```diff
@@ -0,0 +1,202 @@
+	regex      *regexp.Regexp
+}
+
+func newLimitStage(logger log.Logger, config interface{}, registerer prometheus.Registerer) (Stage, error) {
+	cfg := &LimitConfig{}
+
+	err := mapstructure.WeakDecode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	err = validateLimitConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &limitStage{
+		logger:      log.With(logger, "component", "stage", "type", "limit"),
+		cfg:         cfg,
+		dropCount:   getDropCountMetric(registerer),
+		rateLimiter: rate.NewLimiter(rate.Limit(cfg.Rate), cfg.Burst),
+	}
+	return r, nil
+}
+
+func validateLimitConfig(cfg *LimitConfig) error {
+	if cfg == nil ||
+		(cfg.Source == nil && cfg.Expression == nil) {
+		return errors.New(ErrLimitStageEmptyConfig)
+	}
+	if cfg.Value != nil && cfg.Expression != nil {
+		return errors.New(ErrLimitStageInvalidConfig)
+	}
+	if cfg.Expression != nil {
+		expr, err := regexp.Compile(*cfg.Expression)
+		if err != nil {
+			return errors.Errorf(ErrLimitStageInvalidRegex, err)
+		}
+		cfg.regex = expr
+	}
+	if cfg.Rate <= 0 || cfg.Burst <= 0 {
+		return errors.Errorf(ErrLimitStageInvalidRateOrBurst)
+	}
+	return nil
+}
+
+// limitStage applies Label matchers to determine if the include stages should be run
+type limitStage struct {
+	logger      log.Logger
+	cfg         *LimitConfig
+	rateLimiter *rate.Limiter
+	dropCount   *prometheus.CounterVec
+}
+
+func (m *limitStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		for e := range in {
+			if !m.shouldThrottle(e) {
+				out <- e
+				continue
+			}
+		}
+	}()
+	return out
+}
+
+func (m *limitStage) shouldThrottle(e Entry) bool {
+	isMatchTag := m.isMatchTag(e)
+	if !isMatchTag {
+		return false
+	}
+	if m.cfg.Drop {
+		if m.rateLimiter.Allow() {
+			return false
+		}
+		m.dropCount.WithLabelValues(ratelimitDropReason).Inc()
+		return true
+	}
+	_ = m.rateLimiter.Wait(context.Background())
+	return false
+}
+
+// Name implements Stage
+func (m *limitStage) Name() string {
+	return StageTypeLimit
+}
+
+func (m *limitStage) isMatchTag(e Entry) bool {
+	if m.cfg.Source != nil && m.cfg.Expression == nil {
+		if v, ok := e.Extracted[*m.cfg.Source]; ok {
+			if m.cfg.Value == nil {
+				// Found in map, no value set meaning drop if found in map
+				if Debug {
+					level.Debug(m.logger).Log("msg", "line met drop criteria for finding source key in extracted map")
+				}
+			} else {
+				if *m.cfg.Value == v {
+					// Found in map with value set for drop
+					if Debug {
+						level.Debug(m.logger).Log("msg", "line met drop criteria for finding source key in extracted map with value matching desired drop value")
+					}
+				} else {
+					// Value doesn't match, don't drop
+					if Debug {
+						level.Debug(m.logger).Log("msg", fmt.Sprintf("line will not be dropped, source key was found in extracted map but value '%v' did not match desired value '%v'", v, *m.cfg.Value))
+					}
+					return false
+				}
+			}
```

Comment:
This is pretty hard to read, with the multiple levels of indentation.

One thing that'd make it far simpler is having the ability to just run `level.Debug(m.logger).Log(...)` without checking `if Debug` first.